### PR TITLE
Prallel adapters

### PR DESCRIPTION
**Pull request by dsnipe**
More robust refactor for parallel adapter requests.

### CODE REVIEW

File: lib/straight/blockchain_adapters_dispatcher.rb [hunk 1]
```diff
@@ -1,60 +1,67 @@
...
-      wait_values(threads)
-    rescue => e
-      raise e if fault_counter == @step && @list_position == @adapters.size
-      if fault_counter == @step 
-        fault_counter = 0
-        retry
-      end
-      fault_counter += 1
+      execute_in_parallel(block, get_adapters)
+      Timeout::timeout(TIMEOUT, AdaptersTimeoutError) {
```
Comment (AlexanderPavlenko) [inline lib/straight/blockchain_adapters_dispatcher.rb]: I think `Timeout.timeout` looks better. Let's not use `::` for the methods invokation.

File: lib/straight/blockchain_adapters_dispatcher.rb [hunk 2]
```diff
@@ -1,60 +1,67 @@
...
-      loop { break if @result }
-      @result
     end
 
+    def finish_iteration?(attempts)
+      attempts.value == @step
+    end
+
+    def reached_last_adapter?
+      @list_position == @adapter.size
```
Comment (AlexanderPavlenko) [inline lib/straight/blockchain_adapters_dispatcher.rb]: `@adapter`? I do not see variable with such name. Looks like this is not tested and will raise NoMethodError eventually.

File: lib/straight/blockchain_adapters_dispatcher.rb [hunk 3]
```diff
@@ -1,60 +1,67 @@
...
+    def initialize(adapters, tasks_parallel_limit: 2, &block)
+      @list_position        = 0
+      @defer_result         = Concurrent::IVar.new
+      @tasks_parallel_limit = tasks_parallel_limit
+      @step                 = 0
+      @adapters             = adapters
       run_requests(block) if block_given?
     end
 
     def get_adapters
```
Comment (AlexanderPavlenko) [inline lib/straight/blockchain_adapters_dispatcher.rb]: Perhaphs, it's better to mark it as `private`.

File: lib/straight/blockchain_adapters_dispatcher.rb [hunk 4]
```diff
@@ -1,60 +1,67 @@
...
-    def wait_values(threads)
-      threads.each do |t| 
-        Thread.new do
-          if val = t.value
-            @result = val
-            threads.map(&:kill)
-          end
+    def execute_in_parallel(block, adapters)
+      attempts = Concurrent::MVar.new(0)
+      pool = Concurrent::ThreadPoolExecutor.new
```
Comment (AlexanderPavlenko) [inline lib/straight/blockchain_adapters_dispatcher.rb]: Maybe this should be `FixedThreadPool.new(@tasks_parallel_limit)`?
Then you can just put all adapters in a queue and do not mess with the `@list_position` and `@step`.

Comment (dsnipe) [inline lib/straight/blockchain_adapters_dispatcher.rb]: It gives to us different behavior. 
In our case, if one fail, it waits. If others failed, then iterate with next group or finish if result given.
`FixedThreadPool` using free thread as soon as it was released. So it runs request on the next adapter, not awaiting results for current group.

Comment (AlexanderPavlenko) [inline lib/straight/blockchain_adapters_dispatcher.rb]: And what's the point in waiting results of "current group"?

**Pull request merged**